### PR TITLE
Pending state

### PR DESCRIPTION
**Pull request by hkalodner**
Add support for queries against pending state

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 32
           

Complexity increasing per file
==============================
- packages/arb-tx-aggregator/batcher/batcher.go  8
- packages/arb-tx-aggregator/batcher/queue.go  9
         

Complexity decreasing per file
==============================
+ packages/arb-tx-aggregator/web3/codec.go  -4
+ packages/arb-validator-core/ethbridge/chain.go  -3
         

Clones added
============
- packages/arb-validator-core/ethbridge/globalInbox.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6079063&bid=19901458)

File: packages/arb-tx-aggregator/batcher/queue.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+		q.maxNonce = tx.Nonce()
+	}
+	return nil
+}
+
+func (q *txQueue) Empty() bool {
+	return len(q.txes) == 0
+}
+
+func (q *txQueue) Peak() *types.Transaction {
```
Comment (edfelten) [inline packages/arb-tx-aggregator/batcher/queue.go:81]: Should be `Peek`.

File: packages/arb-tx-aggregator/batcher/queue.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/offchainlabs/arbitrum/packages/arb-evm/message"
+	"github.com/offchainlabs/arbitrum/packages/arb-tx-aggregator/snapshot"
+	arbcommon "github.com/offchainlabs/arbitrum/packages/arb-util/common"
+	"log"
+	"math/rand"
+)
+
+// An IntHeap is a min-heap of ints.
```
Comment (edfelten) [inline packages/arb-tx-aggregator/batcher/queue.go:16]: I'm confused by the comments here. The comments seem to imply that this is a priority-queue abstraction (implemented as a heap).  But the code seems to be a simple in-order queue.  I don't see any of the sorting / heapifying operations that a heap would use.

File: packages/arb-tx-aggregator/batcher/queue.go (hunk 1 above)
Comment (hkalodner) [reply packages/arb-tx-aggregator/batcher/queue.go:81]: Fixed

File: packages/arb-tx-aggregator/batcher/queue.go (hunk 2 above)
Comment (hkalodner) [reply packages/arb-tx-aggregator/batcher/queue.go:16]: I updated the comments a bit. The `TxHeap` struct implements the requirements from https://golang.org/pkg/container/heap/ which we use for its heap method. The comments had been copied from the code example located in that documentation

File: packages/arb-tx-aggregator/batcher/batcher.go [hunk 3]
```diff
@@ -69,15 +95,58 @@ func NewBatcher(
...
+					server.Unlock()
+					receipt, err := ethbridge.WaitForReceiptWithResultsSimple(ctx, client, txHash)
+					if err != nil || receipt.Status != 1 {
+						// batch failed
+						log.Fatal("Error submitted batch", err)
+					}
+
+					// batch succeeded
+					server.Lock()
+					server.pendingSentBatches.Remove(server.pendingSentBatches.Front())
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-tx-aggregator/batcher/batcher.go:152]: This is an assumption and I added a comment to that effect

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 9
- Added 31
           

Complexity increasing per file
==============================
- packages/arb-tx-aggregator/batcher/queue.go  9
- packages/arb-tx-aggregator/batcher/batcher.go  8
         

Complexity decreasing per file
==============================
+ packages/arb-tx-aggregator/web3/codec.go  -4
+ packages/arb-validator-core/ethbridge/chain.go  -3
         

Clones added
============
- packages/arb-validator-core/ethbridge/globalInbox.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=6079063&bid=19901458)

**Pull request merged**